### PR TITLE
Fix IndexNotFoundException when deleting partitions

### PR DESCRIPTION
**Pull request by mfussenegger**
When creating and deleting partitions concurrently it could run into a
`IndexNotFoundException`.

The reason was that although we set `IndicesOptions.lenientExpandOpen()`
which allows missing indices, we resolved the `concreteIndices` early
_outside_ of the cluster state update task. Between resolving the
indices and using them a index could disappear, leading to an error.

### CODE REVIEW

File: server/src/test/java/io/crate/integrationtests/PartitionedTableConcurrentIntegrationTest.java [hunk 1]
```diff
@@ -441,4 +443,51 @@ public void testConcurrentPartitionCreationWithColumnCreationTypeMissMatch() thr
...
+                }
+            });
+            thread.start();
+            threads.add(thread);
+        }
+        for (int t = 0; t < numDeleteThreads; t++) {
+            Thread thread = new Thread(() -> {
+                phaser.arriveAndAwaitAdvance();
+                for (int index = 0; index < numOpsPerThread; index++) {
+                    int partition = randomIntBetween(1, 10);
```
Comment (matriv) [inline server/src/test/java/io/crate/integrationtests/PartitionedTableConcurrentIntegrationTest.java:476]: Is this "safe", it can also include partitions that are not created.

File: server/src/test/java/io/crate/integrationtests/PartitionedTableConcurrentIntegrationTest.java [hunk 2]
```diff
@@ -441,4 +443,51 @@ public void testConcurrentPartitionCreationWithColumnCreationTypeMissMatch() thr
...
+
+    @Test
+    public void test_stress_partition_deletion_and_creation() throws Exception {
+        execute(
+            "create table tbl (x int, p int) " +
+            "clustered into 2 shards partitioned by (p) with (number_of_replicas = 0)");
+        int numOpsPerThread = randomIntBetween(100, 300);
+        int numInsertThreads = randomIntBetween(5, 10);
+        int numDeleteThreads = randomIntBetween(5, 10);
+        Phaser phaser = new Phaser(numInsertThreads + numDeleteThreads);
```
Comment (matriv) [inline server/src/test/java/io/crate/integrationtests/PartitionedTableConcurrentIntegrationTest.java:455]: Why is this needed? especially for the inserts? Isn't `register()` required?

**Review by matriv — COMMENTED** [review]
Left a couple of comments, mainly for my understanding of the test.

File: server/src/test/java/io/crate/integrationtests/PartitionedTableConcurrentIntegrationTest.java (hunk 1 above)
Comment (mfussenegger) [reply server/src/test/java/io/crate/integrationtests/PartitionedTableConcurrentIntegrationTest.java:476]: that's part of the idea of the test :)

File: server/src/test/java/io/crate/integrationtests/PartitionedTableConcurrentIntegrationTest.java (hunk 2 above)
Comment (mfussenegger) [reply server/src/test/java/io/crate/integrationtests/PartitionedTableConcurrentIntegrationTest.java:455]: The phaser is used to ensure all threads have started before they run their actual logic.

Not necessarily required, but helps increase the chance that requests execute concurrently

Comment (matriv) [reply server/src/test/java/io/crate/integrationtests/PartitionedTableConcurrentIntegrationTest.java:455]: thx! and you don't need to call `register()` as well? Just had a quick look in `Phaser`'s java doc.

**Review by matriv — APPROVED** [review]
LGTM!
